File: code/5._Huffman_Coding/C++/programming.cpp/Huffman_Coding.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>

struct HuffmanNode
{
    char data;
    int frequency;
    HuffmanNode *left;
    HuffmanNode *right;

    HuffmanNode(char data, int frequency) : data(data), frequency(frequency), left(nullptr), right(nullptr) {}

    bool operator>(const HuffmanNode &other) const
    {
        return frequency > other.frequency;
    }
};

HuffmanNode *buildHuffmanTree(const std::unordered_map<char, int> &freqMap)
{
    std::priority_queue<HuffmanNode, std::vector<HuffmanNode>, std::greater<HuffmanNode>> minHeap;

    for (const auto &entry : freqMap)
    {
        minHeap.push(HuffmanNode(entry.first, entry.second));
    }

    while (minHeap.size() > 1)
    {
        HuffmanNode *left = new HuffmanNode(0, 0);
        HuffmanNode *right = new HuffmanNode(0, 0);

        *left = minHeap.top();
        minHeap.pop();
        *right = minHeap.top();
        minHeap.pop();

        HuffmanNode *merged = new HuffmanNode(0, left->frequency + right->frequency);
        merged->left = left;
        merged->right = right;

        minHeap.push(*merged);
        delete merged;
    }

    return new HuffmanNode(0, minHeap.top().frequency, minHeap.top().left, minHeap.top().right);
}

void generateHuffmanCodes(const HuffmanNode *root, std::string code, std::unordered_map<char, std::string> &huffmanCodes)
{
    if (!root)
    {
        return;
    }

    if (!root->left && !root->right)
    {
        huffmanCodes[root->data] = code;
    }

    generateHuffmanCodes(root->left, code + "0", huffmanCodes);
    generateHuffmanCodes(root->right, code + "1", huffmanCodes);
}

std::string huffmanCoding(const std::string &text)
{
    std::unordered_map<char, int> freqMap;

    for (char c : text)
    {
        freqMap[c]++;
    }

    HuffmanNode *root = buildHuffmanTree(freqMap);
    std::unordered_map<char, std::string> huffmanCodes;

    generateHuffmanCodes(root, "", huffmanCodes);

    std::string encodedText;

    for (char c : text)
    {
        encodedText += huffmanCodes[c];
    }

    return encodedText;
}

int main()
{
    std::string text = "this is an example for huffman encoding";
    std::string encodedText = huffmanCoding(text);
    std::cout << "Huffman Codes:" << std::endl;
    for (const auto &entry : huffmanCodes)
    {
        std::cout << entry.first << ": " << entry.second << std::endl;
    }
    std::cout << "Encoded Text: " << encodedText << std::endl;
    return 0;
}

File: code/5._Huffman_Coding/JavaScript/programming.js/Huffman_Coding.js
class HuffmanNode {
  constructor(data, frequency) {
    this.data = data;
    this.frequency = frequency;
    this.left = null;
    this.right = null;
  }
}

function buildHuffmanTree(freqMap) {
  const minHeap = new PriorityQueue((a, b) => a.frequency - b.frequency);

  for (const [char, freq] of Object.entries(freqMap)) {
    const node = new HuffmanNode(char, freq);
    minHeap.enqueue(node);
  }

  while (minHeap.size() > 1) {
    const left = minHeap.dequeue();
    const right = minHeap.dequeue();

    const merged = new HuffmanNode(null, left.frequency + right.frequency);
    merged.left = left;
    merged.right = right;

    minHeap.enqueue(merged);
  }

  return minHeap.peek();
}

function generateHuffmanCodes(root, code, huffmanCodes) {
  if (root === null) {
    return;
  }

  if (root.data !== null) {
    huffmanCodes[root.data] = code;
  }

  generateHuffmanCodes(root.left, code + "0", huffmanCodes);
  generateHuffmanCodes(root.right, code + "1", huffmanCodes);
}

function huffmanCoding(text) {
  const freqMap = {};

  for (const char of text) {
    freqMap[char] = (freqMap[char] || 0) + 1;
  }

  const root = buildHuffmanTree(freqMap);
  const huffmanCodes = {};
  generateHuffmanCodes(root, "", huffmanCodes);

  let encodedText = "";

  for (const char of text) {
    encodedText += huffmanCodes[char];
  }

  return encodedText;
}

class PriorityQueue {
  constructor(comparator) {
    this.elements = [];
    this.comparator = comparator || ((a, b) => a - b);
  }

  enqueue(element) {
    this.elements.push(element);
    this.bubbleUp();
  }

  dequeue() {
    if (this.isEmpty()) {
      throw new Error("Queue is empty");
    }

    const top = this.elements[0];
    const bottom = this.elements.pop();

    if (!this.isEmpty()) {
      this.elements[0] = bottom;
      this.sinkDown();
    }

    return top;
  }

  peek() {
    if (this.isEmpty()) {
      throw new Error("Queue is empty");
    }

    return this.elements[0];
  }

  size() {
    return this.elements.length;
  }

  isEmpty() {
    return this.elements.length === 0;
  }

  bubbleUp() {
    let index = this.elements.length - 1;

    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);

      if (
        this.comparator(this.elements[index], this.elements[parentIndex]) >= 0
      ) {
        break;
      }

      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  sinkDown() {
    let index = 0;

    while (true) {
      const leftChildIndex = 2 * index + 1;
      const rightChildIndex = 2 * index + 2;
      let smallest = index;

      if (
        leftChildIndex < this.elements.length &&
        this.comparator(
          this.elements[leftChildIndex],
          this.elements[smallest]
        ) < 0
      ) {
        smallest = leftChildIndex;
      }

      if (
        rightChildIndex < this.elements.length &&
        this.comparator(
          this.elements[rightChildIndex],
          this.elements[smallest]
        ) < 0
      ) {
        smallest = rightChildIndex;
      }

      if (smallest === index) {
        break;
      }

      this.swap(index, smallest);
      index = smallest;
    }
  }

  swap(a, b) {
    [this.elements[a], this.elements[b]] = [this.elements[b], this.elements[a]];
  }
}

const text = "this is an example for huffman encoding";
const encodedText = huffmanCoding(text);

console.log("Huffman Codes:");
for (const char in huffmanCodes) {
  console.log(`${char}: ${huffmanCodes[char]}`);
}
console.log(`Encoded Text: ${encodedText}`);
